Patch for gdb-6.2.1
Author: Till Straumann <strauman@slac.stanford.edu> 2003-2005

This patch provides the following features:
 - adds support for NBSD style corefile target for
   powerpc-rtems, i386-rtems and m68k-rtems. NBSD register
   layout was chosen because BFD and GDB have this built in.

   To convert a raw memory image into a suitable core file,
   the 'gencore' utility can be employed. (Comes with the CEXP
   package, I believe.)

 - adds a new 'rtems-remote' target which is an enhanced
   version of 'remote' with the capability to query the
   target for a list of loaded object modules. Taylored
   for supporting a RTEMS/CEXP target with the 'rtems-gdb-stub'
   target agent.

 - bugfix in the powerpc stack unwinder

 - bugfixes for the 'psim' (powerpc simulator) target
   (correctly detect read-only disk; make disk block size
   configurable)

NOTE: To obtain core-file support, GDB must be configured for the
   targets 'powerpc-ssrl-rtems', 'i386-ssrl-rtems' or 'm68k-ssrl-rtems',
   respectively!

To apply this patch, chdir to the GDB-6.2.1 source toplevel directory
and execute
	
		patch -p0 < this_file

(It is always a good idea to try '--dry-run' first)

   Till Straumann, 12/2005

Index: gdb/Makefile.in
===================================================================
RCS file: /afs/slac/g/spear/cvsrep/tools/gdb/gdb/Makefile.in,v
retrieving revision 1.1.1.1
retrieving revision 1.4
diff -c -r1.1.1.1 -r1.4
*** gdb/Makefile.in	10 Nov 2004 18:25:42 -0000	1.1.1.1
--- gdb/Makefile.in	15 Mar 2005 02:10:45 -0000	1.4
***************
*** 535,540 ****
--- 535,541 ----
  	objfiles.c osabi.c observer.c \
  	p-exp.y p-lang.c p-typeprint.c p-valprint.c parse.c printcmd.c \
  	regcache.c reggroups.c remote.c remote-fileio.c \
+ 	remote-rtems.c \
  	scm-exp.c scm-lang.c scm-valprint.c \
  	sentinel-frame.c \
  	serial.c ser-unix.c source.c \
***************
*** 1364,1369 ****
--- 1365,1371 ----
  	i386-tdep.c i386v-nat.c i386-linux-nat.c \
  	i386v4-nat.c i386ly-tdep.c i386-cygwin-tdep.c \
  	i386bsd-nat.c i386bsd-tdep.c i386fbsd-nat.c i386fbsd-tdep.c \
+ 	i386-rtemscore.c \
  	i386nbsd-nat.c i386nbsd-tdep.c i386obsd-nat.c i386obsd-tdep.c \
  	i387-tdep.c \
  	i386-linux-tdep.c i386-nat.c \
***************
*** 1375,1380 ****
--- 1377,1383 ----
  	m68hc11-tdep.c \
  	m68k-tdep.c \
  	m68kbsd-nat.c m68kbsd-tdep.c \
+ 	m68k-rtemscore.c \
  	m88k-tdep.c m88kbsd-nat.c \
  	mcore-tdep.c \
  	mips-linux-nat.c mips-linux-tdep.c \
***************
*** 1388,1393 ****
--- 1391,1397 ----
  	ppc-sysv-tdep.c ppc-linux-nat.c ppc-linux-tdep.c \
  	ppcnbsd-nat.c ppcnbsd-tdep.c \
  	ppcobsd-nat.c ppcobsd-tdep.c \
+ 	ppc32-rtemscore.c \
  	procfs.c \
  	remote-e7000.c \
  	remote-hms.c remote-m32r-sdi.c remote-mips.c \
***************
*** 1939,1944 ****
--- 1943,1949 ----
  	$(reggroups_h) $(regset_h) $(symfile_h) $(symtab_h) $(target_h) \
  	$(value_h) $(dis_asm_h) $(gdb_assert_h) $(gdb_string_h) \
  	$(i386_tdep_h) $(i387_tdep_h)
+ i386-rtemscore.o: i386-rtemscore.c $(defs_h) $(regcache_h) $(i386_tdep_h) $(gregset_h)
  i386v4-nat.o: i386v4-nat.c $(defs_h) $(value_h) $(inferior_h) $(regcache_h) \
  	$(i386_tdep_h) $(i387_tdep_h) $(gregset_h)
  i386v-nat.o: i386v-nat.c $(defs_h) $(frame_h) $(inferior_h) $(language_h) \
***************
*** 2075,2080 ****
--- 2080,2086 ----
  	$(gdbcore_h) $(value_h) $(gdb_string_h) $(gdb_assert_h)	\
  	$(inferior_h) $(regcache_h) $(arch_utils_h) $(osabi_h) $(dis_asm_h) \
  	$(m68k_tdep_h) $(gregset_h)
+ m68k-rtemscore.o: m68k-rtemscore.c $(defs_h) $(regcache_h) $(m68k_tdep_h) $(gregset_h)
  m88kbsd-nat.o: m88kbsd-nat.c $(defs_h) $(inferior_h) $(regcache_h) \
  	$(gdb_assert_h) $(m88k_tdep_h)
  m88k-tdep.o: m88k-tdep.c $(defs_h) $(arch_utils_h) $(dis_asm_h) $(frame_h) \
***************
*** 2208,2213 ****
--- 2214,2220 ----
  	$(symtab_h) $(target_h) $(gdbcore_h) $(gdbcmd_h) $(symfile_h) \
  	$(objfiles_h) $(regcache_h) $(value_h) $(osabi_h) $(regset_h) \
  	$(solib_svr4_h) $(ppc_tdep_h) $(trad_frame_h) $(frame_unwind_h)
+ ppc32-rtemscore.o: ppc32-rtemscore.c $(defs_h) $(regcache_h) $(ppc_tdep_h) $(gregset_h)
  ppcnbsd-nat.o: ppcnbsd-nat.c $(defs_h) $(inferior_h) $(ppc_tdep_h) \
  	$(ppcnbsd_tdep_h) $(gdb_assert_h)
  ppcnbsd-tdep.o: ppcnbsd-tdep.c $(defs_h) $(gdbcore_h) $(regcache_h) \
***************
*** 2250,2256 ****
  reggroups.o: reggroups.c $(defs_h) $(reggroups_h) $(gdbtypes_h) \
  	$(gdb_assert_h) $(regcache_h) $(command_h) $(gdbcmd_h)
  regset.o: regset.c $(defs_h) $(regset_h) $(gdb_assert_h)
! remote.o: remote.c $(defs_h) $(gdb_string_h) $(inferior_h) $(bfd_h) \
  	$(symfile_h) $(target_h) $(gdbcmd_h) $(objfiles_h) $(gdb_stabs_h) \
  	$(gdbthread_h) $(remote_h) $(regcache_h) $(value_h) $(gdb_assert_h) \
  	$(event_loop_h) $(event_top_h) $(inf_loop_h) $(serial_h) \
--- 2257,2263 ----
  reggroups.o: reggroups.c $(defs_h) $(reggroups_h) $(gdbtypes_h) \
  	$(gdb_assert_h) $(regcache_h) $(command_h) $(gdbcmd_h)
  regset.o: regset.c $(defs_h) $(regset_h) $(gdb_assert_h)
! remote.o: remote.c remote-rtems.c $(defs_h) $(gdb_string_h) $(inferior_h) $(bfd_h) \
  	$(symfile_h) $(target_h) $(gdbcmd_h) $(objfiles_h) $(gdb_stabs_h) \
  	$(gdbthread_h) $(remote_h) $(regcache_h) $(value_h) $(gdb_assert_h) \
  	$(event_loop_h) $(event_top_h) $(inf_loop_h) $(serial_h) \
Index: gdb/configure.tgt
===================================================================
RCS file: /afs/slac/g/spear/cvsrep/tools/gdb/gdb/configure.tgt,v
retrieving revision 1.1.1.1
retrieving revision 1.2
diff -c -r1.1.1.1 -r1.2
*** gdb/configure.tgt	10 Nov 2004 18:25:43 -0000	1.1.1.1
--- gdb/configure.tgt	10 Nov 2004 23:12:39 -0000	1.2
***************
*** 94,99 ****
--- 94,100 ----
  			configdirs="${configdirs} nlm" ;;
  i[34567]86-*-cygwin*)	gdb_target=cygwin  ;;
  i[34567]86-*-vxworks*)	gdb_target=vxworks ;;
+ i[34567]86-ssrl-rtems*) gdb_target=i386-ssrlrtems ;;
  i[34567]86-*-*)		gdb_target=i386 ;;
  
  ia64-*-aix*)		gdb_target=aix ;;
***************
*** 112,117 ****
--- 113,119 ----
  m68*-*-aout*)		gdb_target=monitor ;;
  m68*-*-coff*)		gdb_target=monitor ;;
  m68*-*-elf*)		gdb_target=monitor ;;
+ m68*-ssrl-rtems*)   gdb_target=m68k-monitor-ssrlrtems ;;
  m68*-*-linux*)		gdb_target=linux
  			build_gdbserver=yes
  			;;
***************
*** 147,152 ****
--- 149,155 ----
  			;;
  powerpc64-*-linux*)	gdb_target=linux ;;
  powerpc-*-vxworks*)	gdb_target=vxworks ;;
+ powerpc-ssrl-rtems*) gdb_target=ppc-ssrlrtems;;
  powerpc*-*-*)		if test -f ../sim/ppc/Makefile; then
  			  gdb_target=ppc-sim
  			else
Index: gdb/i386-rtemscore.c
===================================================================
RCS file: gdb/i386-rtemscore.c
diff -N gdb/i386-rtemscore.c
*** /dev/null	1 Jan 1970 00:00:00 -0000
--- gdb/i386-rtemscore.c	10 Nov 2004 23:12:39 -0000	1.1
***************
*** 0 ****
--- 1,35 ----
+ /* Prototypes for supply_gregset etc. */
+  
+ #include "defs.h"
+ #include "gregset.h"
+ #include "i386-tdep.h"
+ #include "regcache.h"
+ 
+ void
+ supply_gregset (gdb_gregset_t *gregsetp)
+ {
+ unsigned *regs = (unsigned *)gregsetp;
+ struct   gdbarch_tdep *tdep = gdbarch_tdep (current_gdbarch); 
+ 
+ 	supply_register( I386_EAX_REGNUM,      regs ); regs++;
+ 	supply_register( I386_ECX_REGNUM,      regs ); regs++;
+ 	supply_register( I386_EDX_REGNUM,      regs ); regs++;
+ 	supply_register( I386_EBX_REGNUM,      regs ); regs++;
+ 	supply_register( I386_ESP_REGNUM,      regs ); regs++;
+ 	supply_register( I386_EBP_REGNUM,      regs ); regs++;
+ 	supply_register( I386_ESI_REGNUM,      regs ); regs++;
+ 	supply_register( I386_EDI_REGNUM,      regs ); regs++;
+ 	supply_register( I386_EIP_REGNUM,      regs ); regs++;
+ 	supply_register( I386_EFLAGS_REGNUM,   regs ); regs++;
+ 	supply_register( I386_CS_REGNUM,       regs ); regs++;
+ 	supply_register( I386_SS_REGNUM,       regs ); regs++;
+ 	supply_register( I386_DS_REGNUM,       regs ); regs++;
+ 	supply_register( I386_ES_REGNUM,       regs ); regs++;
+ 	supply_register( I386_FS_REGNUM,       regs ); regs++;
+ 	supply_register( I386_GS_REGNUM,       regs ); regs++;
+ }
+ 
+ void
+ supply_fpregset (gdb_fpregset_t *gregsetp)
+ {
+ }
Index: gdb/m68k-rtemscore.c
===================================================================
RCS file: gdb/m68k-rtemscore.c
diff -N gdb/m68k-rtemscore.c
*** /dev/null	1 Jan 1970 00:00:00 -0000
--- gdb/m68k-rtemscore.c	10 Nov 2004 18:30:59 -0000	1.1
***************
*** 0 ****
--- 1,30 ----
+ /* Prototypes for supply_gregset etc. */
+ 
+ #include "defs.h"
+ #include "gregset.h"
+ #include "m68k-tdep.h"
+ #include "regcache.h"
+ 
+ void
+ supply_gregset (gdb_gregset_t *gregsetp)
+ {
+ unsigned *regs = (unsigned *)gregsetp;
+ int i;
+ 
+ 	for ( i=0; i<8; i++, regs++ )
+   		supply_register (M68K_D0_REGNUM + i, (char*)regs);
+ 
+ 	for ( i=0; i<6; i++, regs++ )
+   		supply_register (M68K_A0_REGNUM + i, (char*)regs);
+ 
+ 	supply_register (M68K_FP_REGNUM, (char*)regs); regs++;
+ 	supply_register (M68K_SP_REGNUM, (char*)regs); regs++;
+ 
+ 	supply_register (M68K_PS_REGNUM, (char*)regs); regs++;
+ 	supply_register (M68K_PC_REGNUM, (char*)regs); regs++;
+ }
+ 
+ void
+ supply_fpregset (gdb_fpregset_t *gregsetp)
+ {
+ }
Index: gdb/ppc32-rtemscore.c
===================================================================
RCS file: gdb/ppc32-rtemscore.c
diff -N gdb/ppc32-rtemscore.c
*** /dev/null	1 Jan 1970 00:00:00 -0000
--- gdb/ppc32-rtemscore.c	10 Nov 2004 18:30:59 -0000	1.1
***************
*** 0 ****
--- 1,33 ----
+ /* Prototypes for supply_gregset etc. */
+  
+ #include "defs.h"
+ #include "gregset.h"
+ #include "ppc-tdep.h"
+ #include "regcache.h"
+ 
+ void
+ supply_gregset (gdb_gregset_t *gregsetp)
+ {
+ unsigned *regs = (unsigned *)gregsetp;
+ int		 regi;
+ struct   gdbarch_tdep *tdep = gdbarch_tdep (current_gdbarch); 
+ 
+   for (regi = 0; regi < 32; regi++)
+     supply_register (regi, regs + regi);
+ 
+   supply_register (tdep->ppc_lr_regnum,  regs + 32 );
+   supply_register (tdep->ppc_cr_regnum,  regs + 33 );
+   supply_register (tdep->ppc_xer_regnum, regs + 34 );
+   supply_register (tdep->ppc_ctr_regnum, regs + 35 );
+   supply_register (PC_REGNUM,            regs + 36 );
+ 
+   /* Netbsd doesn't define this; the SVGM bsp puts it there
+    * and 'gencore' writes our extensions :-)
+    */
+   supply_register (tdep->ppc_ps_regnum,  regs + 37 );
+ }
+ 
+ void
+ supply_fpregset (gdb_fpregset_t *gregsetp)
+ {
+ }
Index: gdb/remote-rtems.c
===================================================================
RCS file: gdb/remote-rtems.c
diff -N gdb/remote-rtems.c
*** /dev/null	1 Jan 1970 00:00:00 -0000
--- gdb/remote-rtems.c	30 Nov 2005 20:28:51 -0000	1.5
***************
*** 0 ****
--- 1,616 ----
+ static char ** remote_file_list(void);
+ static struct section_addr_info *remote_section_list(char *);
+ static void    init_rtems_ops(void);
+ static void    discard_remote_objfile(struct objfile *of);
+ static void    add_remote_objfiles(char **fl, int);
+ static char ** split_cmdline (char *args, int *pnargs);
+ 
+ static struct target_ops rtems_ops;
+ 
+ /* call either with 'os' or with 'os==NULL' but a valid
+  * bfd/section pair...
+  */
+ static int
+ compare_section (struct obj_section *os, bfd *abfd, asection *s)
+ {
+   struct remote_state *rs = get_remote_state ();
+   unsigned long host_crc, target_crc;
+   char *tmp;
+   char *sectdata;
+   char *buf = alloca (rs->remote_packet_size);
+   bfd_size_type size;
+   bfd_vma vma;
+ 
+   if ( os )
+ 	{
+ 	  s = os->the_bfd_section;
+ 	  abfd = os->objfile->obfd;
+ 	}
+ 
+   if (!(s->flags & SEC_LOAD))
+ 	internal_error(__FILE__, __LINE__, "section '%s' not loadable", bfd_get_section_name( abfd, s));
+ 
+   size = bfd_get_section_size (s);
+   if (size == 0)
+ 	return 0;		/* skip zero-length section */
+ 
+   vma  = bfd_get_section_vma (abfd, s);
+   if (os)
+ 	vma += os->addr;
+   /* FIXME: assumes lma can fit into long */
+   sprintf (buf, "qCRC:%lx,%lx", (long) vma, (long) size);
+   putpkt (buf);
+ 
+   /* be clever; compute the host_crc before waiting for target reply */
+   sectdata = xmalloc (size);
+   bfd_get_section_contents (abfd, s, sectdata, 0, size);
+   host_crc = crc32 ((unsigned char *) sectdata, size, 0xffffffff);
+   xfree (sectdata);
+ 
+   getpkt (buf, (rs->remote_packet_size), 0);
+   if (buf[0] == 'E')
+ 	error ("target memory fault, section %s, range 0x%s -- 0x%s",
+ 	       bfd_get_section_name(abfd, s), paddr (vma), paddr (vma + size));
+   if (buf[0] != 'C')
+ 	{
+ 	  warning ("remote target does not support CRC operation - skipping...");
+ 	  return 0;
+ 	}
+ 
+   for (target_crc = 0, tmp = &buf[1]; *tmp; tmp++)
+ 	target_crc = target_crc * 16 + fromhex (*tmp);
+ 
+   return target_crc - host_crc;
+ }
+ 
+ 
+ struct saa_ {
+ 	bfd                     *sbfd;
+ 	int                      ftty;
+ 	struct section_addr_info *sa;
+ 	struct objfile			 *of;
+ };
+ 
+ static int
+ sfa(void *arg)
+ {
+ struct saa_    *sfaa = arg;
+ 
+   sfaa->of = symbol_file_add_from_bfd (sfaa->sbfd, sfaa->ftty, sfaa->sa, 0, 0);
+ #if 0
+ /* This doesn't work -- on the target, the section contents
+  * are relocated and hence differing from the ones in the
+  * file :-(
+  */
+   if ( compare_section ( &of->sections[of->sect_index_text], 0, 0 ) )
+ 	error("%s: '.text' section of object file doesn't match the target's -- do GDB and the target use the same file?", bfd_get_filename(of->obfd));
+ #endif
+   return -1;
+ }
+ 
+ static struct objfile *
+ caught_symbol_file_add_from_bfd( bfd *sbfd, int from_tty, struct section_addr_info *sa)
+ {
+   struct saa_ sfaa;
+   sfaa.sbfd = sbfd;
+   sfaa.ftty = from_tty;
+   sfaa.sa   = sa;
+   sfaa.of   = 0;
+   catch_errors( sfa, &sfaa, "Omitting -- ", RETURN_MASK_ALL );
+   return sfaa.of;
+ }
+ 
+ struct sbd_ {
+ 	char	*name;
+ 	bfd		*sbfd;
+ };
+ 
+ static int
+ sbdo(void *arg)
+ {
+ struct sbd_ *sbda = arg;
+ 	sbda->sbfd = symfile_bfd_open (sbda->name);
+ 	return -1;
+ }
+ 
+ static bfd *
+ caught_symfile_bfd_open(char *name)
+ {
+ 	struct sbd_ sbda;
+ 	sbda.name = name;
+ 	sbda.sbfd = 0;
+ 	catch_errors( sbdo, &sbda, "Omitting -- ", RETURN_MASK_ALL );
+ 	return sbda.sbfd;
+ }
+ 
+ static char*
+ rtems_pid_to_str (ptid_t ptid)
+ {
+   static char buf[30];
+ 
+   sprintf(buf, "Thread %8x", PIDGET (ptid));
+   return buf;
+ }
+ 
+ static void rtems_remote_open(char *name, int from_tty)
+ {
+ extern bfd *exec_bfd;
+ 
+   discard_remote_objfile(0);
+   remote_open_1 (name, from_tty, &rtems_ops, 0, 0);
+   if ( exec_bfd )
+ 	{
+ 	  asection *s;
+ 	  if ( !(s=bfd_get_section_by_name (exec_bfd, ".text")) ) {
+ 		error("No '.text' section found in executable");
+ 	  }
+ 	  if ( compare_section ( 0, exec_bfd, s ) )
+ 		warning("%s: '.text' section of executable file doesn't match the target's -- do GDB and the target use the same file?", bfd_get_filename(exec_bfd));
+ 	}
+   add_remote_objfiles(0,from_tty);
+ }
+ 
+ static void rtems_files_info(struct target_ops *t)
+ {
+ struct objfile *of;
+ 
+   ALL_OBJFILES(of)
+ 	{
+ 	  if ( of->obfd )
+ 		print_section_info(t, of->obfd);
+ 	  else
+ 		{
+ 		  puts_filtered(of->name); puts_filtered("\n");
+ 		}
+ 	}
+ }
+ 
+ static char **
+ remote_file_list(void)
+ {
+   struct remote_state      *rs = get_remote_state ();
+   char                     *buf = alloca (rs->remote_packet_size);
+   char                     *bufp, *n;
+   int                      i, num_entries, chunks, addr;
+   char                     **file_list = 0;
+ 
+   if (remote_desc == 0)		/* paranoia */
+     error ("need connection to the remote target for retrieving a file list.");
+ 
+   putpkt ("qfCexpFileList");
+   bufp = buf;
+   getpkt (bufp, (rs->remote_packet_size), 0);
+   if (bufp[0] != '\0') {
+     /* packet recognized */
+ 	{
+ 	  i = 0;
+ 	  chunks = 0;
+ 	  while (*bufp++ == 'm')	/* reply contains one 'text_addr ','file_name' pair */
+ 		{
+ 		  if ( i <= chunks )
+ 			{
+ 			  chunks += 15;
+ 			  file_list = xrealloc (file_list, chunks * sizeof (*file_list));
+ 		    }
+ 		  addr = strtol (bufp, &bufp, 16); /* currently unused */
+ 		  if ( bufp == buf+1 || ',' != *bufp++ )
+ 			{
+ 			  xfree (file_list); file_list = 0; i = chunks = 0;  /* paranoia -- error shouldn't return */
+ 			  error ("ill formated file list packet");
+ 		 	  continue;
+ 			}
+ 		  /* buf is guaranteed to be null terminated */
+ 		  n = file_list[i] = strdup (bufp);
+ 		  i++;
+ 		  make_cleanup (xfree, n);
+ 	      putpkt ("qsCexpFileList");
+ 	      bufp = buf;
+ 	      getpkt (bufp, (rs->remote_packet_size), 0);
+ 	    }
+ 	  if ( i<= chunks )
+ 		file_list = xrealloc (file_list, (i + 1) * sizeof (*file_list));
+ 	  file_list[i++] = 0;
+ 	  make_cleanup (xfree, file_list);
+ 	  return file_list;	/* done */
+ 	}
+   }
+   return 0;
+ }
+ 
+ /* allocates and populates a section_addr_info structure
+  * with name/addr info obtained from the target.
+  * returned array and 'name' fields must be cleaned up
+  * by the caller
+  */
+ 
+ static struct section_addr_info *
+ remote_section_list(char *filename)
+ {
+   struct remote_state      *rs = get_remote_state ();
+   char                     *buf = alloca (rs->remote_packet_size);
+   char                     *bufp, *n;
+   int                      i, num_entries, chunks;
+   struct section_addr_info *section_addrs = NULL;
+   struct sal_ {
+     int  addr;
+ 	char *name;
+   } *sal = 0;
+ 
+   if (remote_desc == 0)		/* paranoia */
+     error ("need connection to the remote target for retrieving a section list.");
+ 
+   gdb_assert (strlen(filename) < rs->remote_packet_size + 20);
+   sprintf (buf,"qfCexpSectionList,%s",filename);
+   putpkt (buf);
+   bufp = buf;
+   getpkt (bufp, (rs->remote_packet_size), 0);
+   if (bufp[0] != '\0') {
+     /* packet recognized */
+ 	{
+ 	  i = 0;
+ 	  chunks = 0;
+ 	  while (*bufp++ == 'm')	/* reply contains one section 'addr ',' name' pair */
+ 		{
+ 		  if ( i <= chunks )
+ 			{
+ 			  chunks += 15;
+ 			  sal = xrealloc (sal, chunks * sizeof (*sal));
+ 		    }
+ 		  sal[i].addr = strtol (bufp, &bufp, 16);
+ 		  if ( bufp == buf+1 || ',' != *bufp++ )
+ 			{
+ 			  xfree (sal); sal = 0; i = chunks = 0;  /* paranoia -- error shouldn't return */
+ 			  error ("ill formated section list packet");
+ 		 	  continue;
+ 			}
+ 		  /* buf is guaranteed to be null terminated */
+ 		  n = sal[i].name = strdup (bufp);
+ 		  i++;
+ 		  make_cleanup (xfree, n);
+ 	      putpkt ("qsCexpSectionList");
+ 	      bufp = buf;
+ 	      getpkt (bufp, (rs->remote_packet_size), 0);
+ 	    }
+ 	  num_entries = i;
+ 	  section_addrs = alloc_section_addr_info (num_entries);
+ 	  make_cleanup (xfree, section_addrs);
+ 	  for (i = 0; i<num_entries; i++)
+ 		{
+ 		  section_addrs->other[i].addr = sal[i].addr;
+ 		  section_addrs->other[i].name = sal[i].name;
+ 		}
+ 	  xfree (sal);
+ 	  if ( 'E' == buf[0] )
+ 		{
+ 			int errn;
+ 			unpack_byte(buf+1,&errn);
+ 			error ("Target Error: %s\n", strerror (errn));
+ 		}
+ 	  return section_addrs;	/* done */
+ 	}
+   }
+   return 0;
+ }
+ 
+ 
+ static void
+ discard_remote_objfile (struct objfile *objfile)
+ /* throw away old symbols */
+ {
+   struct objfile *temp;
+ 
+   if ( objfile )
+ 	{
+ 	  if ( symfile_objfile != objfile )
+ 		free_objfile (objfile);
+ 	}
+   else
+ 	{
+   ALL_OBJFILES_SAFE (objfile, temp)
+ 	{
+ 	  if ( symfile_objfile != objfile )
+ 		free_objfile (objfile);
+ 	}
+ 	}
+ 
+   clear_symtab_users ();
+ 
+   if (remote_debug)
+  	{
+ 	  ALL_OBJFILES(objfile)
+ 	  printf_filtered("still have: '%s'\n", objfile->name);
+ 	}
+ }
+ 
+ 
+ static void
+ add_remote_objfiles (char **fl, int from_tty)
+ {
+   struct cleanup           *cln = make_cleanup (null_cleanup, 0);
+   struct section_addr_info *sa;
+   int            i,j;
+ 
+   if ( !fl )
+ 	fl = remote_file_list ();
+   if (remote_debug)
+ 	printf_filtered ("Target files:\n");
+   for ( i=0; fl && fl[i]; i++ )
+ 	{
+ 	  struct cleanup *cln1 = 0;
+ 	  struct objfile *of;
+ 	  bfd            *sbfd;
+ 
+ 	  if (remote_debug)
+ 	    printf_filtered ("  %-40sSection Info:\n",fl[i]);	
+ 
+ 	  if ( (sbfd = caught_symfile_bfd_open(fl[i])) )
+ 		{
+ 		  sa = remote_section_list (fl[i]);
+ 		  if ( !sa )
+ 			{
+ 			  bfd_close(sbfd);
+ 			  continue;
+ 			} 
+ 		  cln1 = make_cleanup(null_cleanup, 0);
+ 
+ 		  if (remote_debug)
+ 			{
+ 			  for (j = 0; j < sa->num_sections; j++)
+ 				{
+ 				  printf_filtered ("    0x%08lx: %s\n", sa->other[j].addr, sa->other[j].name);
+ 				}
+ 			}
+   
+ 		  /* Warn about common symbols */
+ 		  if ( (of = caught_symbol_file_add_from_bfd ( sbfd, from_tty, sa ))
+ 			&& (HAS_SYMS & bfd_get_file_flags( of->obfd )) )
+ 			{
+ 			  int     i = bfd_get_symtab_upper_bound( of->obfd );
+ 			  if ( i > 0 )
+ 				{
+ 			  	  asymbol **syms = (asymbol**)xmalloc(i);
+ 				  make_cleanup(xfree, syms);
+ 				  if ( (i = bfd_canonicalize_symtab( of->obfd, syms )) > 0 )
+ 					{
+ 					  int k,w = 0;
+ 					  for ( k=0; k < i; k++ )
+ 						{
+ 						  if ( bfd_is_com_section(bfd_get_section(syms[k])) )
+ 							{
+ 							  if ( !w )
+ 								{
+ 							  	  warning("COMMON symbols found:");
+ 								  w = 1;
+ 								}
+ 							  printf_filtered(" %s\n", bfd_asymbol_name(syms[k]));
+ 							}
+ 						}
+ 					  if ( w )
+ 						{
+ 						  warning("Unable to determine address of COMMONs -- compile with -fno-common or DO NOT USE the mentioned symbols");
+ 						}
+ 					}
+ 				}
+ 			}
+ 		  do_cleanups(cln1);
+ 		}
+  	}
+   reinit_frame_cache ();
+   do_cleanups (cln);
+ }
+ 
+ static char **
+ split_cmdline (char *args, int *pnargs)
+ {
+   char  *blnk = args;
+   char **rval = 0;
+   int   nargs = 1; /* extra for arg[0] */
+ 
+   *pnargs = 0;
+   while (blnk)
+ 	{
+ 	  while ( isspace(*blnk) )
+ 		*blnk++ = 0;
+ 	  if ( *blnk )
+ 		{
+ 		  if ( !rval )
+ 			{
+ 			  *pnargs = nargs;
+ 			  blnk    = strdup(blnk);
+ 			  rval    = xcalloc(1,sizeof(*rval));
+ 			}
+ 		  nargs++;
+ 		  rval = xrealloc(rval, nargs*sizeof(*rval));
+ 		  rval[*pnargs] = blnk;
+ 		  *pnargs = nargs;
+ 		  blnk = strchr(blnk,' ');
+ 		}
+ 	  else
+ 		{
+ 		  blnk = 0;
+ 		}
+ 	}
+   return rval;
+ }
+ 
+ static void
+ rtems_kill()
+ {
+   remote_kill();
+   pop_target();
+ }
+ 
+ static void
+ rtems_mourn(void)
+ {
+   remote_mourn_1 (&rtems_ops);
+ }
+ 
+ static void
+ rtems_disconnect(char *args, int from_tty)
+ {
+ 	remote_disconnect(args, from_tty);
+ 	pop_target();
+ }
+ 
+ static void
+ rtems_detach(char *args, int from_tty)
+ {
+ 	remote_detach(args, from_tty);
+ 	pop_target();
+ }
+ 
+ static void
+ rtems_load_file_command (char *args, int from_tty)
+ {
+   struct remote_state      *rs = get_remote_state ();
+   char                     *buf = alloca (rs->remote_packet_size);
+   int                      nargs, ch;
+   int                      unload = 0, refresh = 0;
+   char                     **aa   = split_cmdline(args, &nargs);
+   char                     *fnam, *bufp, *errmsg = 0;
+ 
+   if (aa)
+ 	{
+ 	  make_cleanup(xfree, aa[1]);
+ 	  make_cleanup(xfree, aa);
+ 	}
+ 
+   optind = 0;
+   while ( (ch = getopt(nargs, aa, "u")) > 0 )
+ 	{
+ 	  switch (ch)
+ 		{
+ 		  default: error("unknown option: '%c'",ch);
+ 		  case 'u': unload = 1; break;
+ 		}
+ 	}
+ 
+   if ( optind >= nargs )
+ 	error("need a filename argument");
+ 
+   fnam = aa[optind];
+ 
+   if (remote_desc == 0)		/* paranoia */
+     error ("need connection to the remote target for loading an object file.");
+ 
+   gdb_assert (strlen(fnam) < rs->remote_packet_size + 25);
+   sprintf (buf,"qCexp%s,%s",unload?"Unld":"Load",fnam);
+   putpkt (buf);
+   bufp = buf;
+   getpkt (bufp, (rs->remote_packet_size), 0);
+   if (bufp[0] != '\0')
+ 	{
+ 	  if ( !strcmp("OK", bufp) )
+ 		{
+ 		  refresh = 1;
+ 		}
+ 	  else if ( !strcmp("E10", bufp) )
+ 		{
+ 		  errmsg="Unable to unload -- module still in use";
+ 		}
+ 	  else if ( !strcmp("E02", bufp) )
+ 		{
+ 		  refresh = 1;
+ 		  errmsg="File not found - check PATH on target";
+ 		}
+ 	  else
+ 		{
+ 		  refresh = 1;
+ 		  errmsg="Unknown error";
+ 		}
+ 	}
+   else
+ 	{
+ 	  errmsg="qCexpLoad/qCexpUnld not supported by target";
+ 	}
+ 
+   if (refresh)
+ 	{
+ 	  discard_remote_objfile(0);
+ 	  add_remote_objfiles(0, from_tty);
+ 	}
+ 
+   if (errmsg)
+ 	error("%s",errmsg);
+ }
+ 
+ struct cmd_list_element *rtems_cmd_list = 0;
+ 
+ static void rtems_refresh_files_command (char *args, int from_tty)
+ {
+   discard_remote_objfile(0);
+   add_remote_objfiles(0, from_tty);
+ }
+ 
+ static void rtems_load_redirect (char *args, int from_tty)
+ {
+   error("Please use the \"rtems load\" command");
+ }
+ 
+ static void rtems_cmd (char *args, int from_tty)
+ {
+   printf_unfiltered("\"rtems\" must be followed by the name of a RTEMS specific command.\n");
+   help_list(rtems_cmd_list, "rtems ", -1, gdb_stdout);
+ }
+ 
+ static void init_rtems_cmds ()
+ {
+   add_prefix_cmd ("rtems", all_classes, rtems_cmd,
+ 		  "RTEMS specific commands",
+ 		  &rtems_cmd_list, "rtems ",
+ 		  0/*allow-unknown*/, &cmdlist);
+   add_cmd ("sync-objs", class_files, rtems_refresh_files_command,
+ 	   "Synchronize gdb's file/section data with what's current on the target",
+ 	   &rtems_cmd_list);
+   add_cmd ("load", class_files, rtems_load_file_command,
+        "rtems load [-u] <obj_file_name>\n"
+ 	   "Load/link unload/unlink (use [-u] option) object files on the target\n"
+        "NOTE: the object file must be found in GDB 'path' AND target 'PATH'",
+ 	   &rtems_cmd_list);
+ }
+ 
+ 
+ static void 
+ init_rtems_ops ()
+ {
+   rtems_ops						= remote_ops;
+ 
+   if ( rtems_ops.to_open != remote_open )
+ 	{
+ 	  internal_error(__FILE__, __LINE__, "'rtems' target MUST be initialized AFTER 'remote'");
+ 	}
+   /* inherit from 'remote' */
+   memcpy(&rtems_ops, &remote_ops, sizeof(rtems_ops));
+   /* need to have 'remote' in the name for the compare-sections command to work */
+   rtems_ops.to_shortname    = "rtems-remote";
+   rtems_ops.to_longname     = "rtems target via 'remote' protocol";
+   rtems_ops.to_doc          = "rtems target; specify link: <serial_dev> | <host>:<port>";
+   rtems_ops.to_open         	= rtems_remote_open;
+   rtems_ops.to_files_info   	= rtems_files_info;
+   rtems_ops.to_pid_to_str   	= rtems_pid_to_str;
+   rtems_ops.to_load				= rtems_load_redirect;
+ /*  rtems_ops.to_kill			= rtems_kill; */
+   rtems_ops.to_mourn_inferior	= rtems_mourn;
+   rtems_ops.to_disconnect		= rtems_disconnect;
+   rtems_ops.to_detach			= rtems_detach;
+   rtems_ops.to_magic        	= OPS_MAGIC;
+ }
+ 
+ void
+ _initialize_remote_rtems (void)
+ {
+   /* TILL a temporary hack...; should be queried from the
+    * target. Also, it seems that we could avoid a lot of grief
+    * by having this for any architecture. The red-zone makes
+    * our life a LOT easier (no stack switching needed). The
+    * architecture needs a 'frame_align' method, though but this
+    * should be straightforward to provide (unless I miss something)
+    */
+   set_gdbarch_frame_red_zone_size( current_gdbarch, 4096 );
+   set_main_name("_Thread_Handler");
+ 
+   init_rtems_ops();
+   add_target (&rtems_ops);
+   init_rtems_cmds ();
+ }
Index: gdb/remote.c
===================================================================
RCS file: /afs/slac/g/spear/cvsrep/tools/gdb/gdb/remote.c,v
retrieving revision 1.1.1.1
diff -c -r1.1.1.1 remote.c
*** gdb/remote.c	10 Nov 2004 18:25:48 -0000	1.1.1.1
--- gdb/remote.c	28 Sep 2005 20:52:49 -0000
***************
*** 3192,3197 ****
--- 3192,3199 ----
  	fprintf_unfiltered (gdb_stdlog,
  			    "Bad register packet; fetching a new packet\n");
        getpkt (buf, (rs->remote_packet_size), 0);
+ 	  if (buf[0] == 0)
+ 		break;
      }
  
    /* Reply describes registers byte by byte, each byte encoded as two
***************
*** 4721,4727 ****
    if (!exec_bfd)
      error ("command cannot be used without an exec file");
    if (!current_target.to_shortname ||
!       strcmp (current_target.to_shortname, "remote") != 0)
      error ("command can only be used with remote target");
  
    for (s = exec_bfd->sections; s; s = s->next)
--- 4723,4729 ----
    if (!exec_bfd)
      error ("command cannot be used without an exec file");
    if (!current_target.to_shortname ||
!       (strstr (current_target.to_shortname, "remote") == 0) )
      error ("command can only be used with remote target");
  
    for (s = exec_bfd->sections; s; s = s->next)
***************
*** 5574,5576 ****
--- 5576,5585 ----
    /* Eventually initialize fileio.  See fileio.c */
    initialize_remote_fileio (remote_set_cmdlist, remote_show_cmdlist);
  }
+ 
+ /* need to introduce this name here ( to be
+  * seen by sed -e 's/^_initialize_.../'
+  */
+ void
+ _initialize_remote_rtems (void);
+ #include "remote-rtems.c"
Index: gdb/rs6000-tdep.c
===================================================================
RCS file: /afs/slac/g/spear/cvsrep/tools/gdb/gdb/rs6000-tdep.c,v
retrieving revision 1.1.1.1
retrieving revision 1.3
diff -c -r1.1.1.1 -r1.3
*** gdb/rs6000-tdep.c	10 Nov 2004 18:25:48 -0000	1.1.1.1
--- gdb/rs6000-tdep.c	12 Nov 2005 00:43:01 -0000	1.3
***************
*** 849,855 ****
  	  continue;
  
  	}
!       else if (lr_reg != -1 &&
  	       /* std Rx, NUM(r1) || stdu Rx, NUM(r1) */
  	       (((op & 0xffff0000) == (lr_reg | 0xf8010000)) ||
  		/* stw Rx, NUM(r1) */
--- 849,855 ----
  	  continue;
  
  	}
!       else if (lr_reg != -1 && fdata->nosavedpc &&
  	       /* std Rx, NUM(r1) || stdu Rx, NUM(r1) */
  	       (((op & 0xffff0000) == (lr_reg | 0xf8010000)) ||
  		/* stw Rx, NUM(r1) */
***************
*** 894,900 ****
  	  continue;
  
  	}
!       else if (op == 0x48000004)
  	{			/* b .+4 (xlc) */
  	  break;
  
--- 894,906 ----
  	  continue;
  
  	}
!  	  else if ( (( op & 0xfe80ffff ) == 0x4280004) || (( op & 0xfe80ffff ) == 0x42800005 && lr_reg > -1 ) )
!  	{			/* bcl .+4 used in -fpic */
!  	
!  	  continue;
!  
!  	}
!        else if (op == 0x48000004)
  	{			/* b .+4 (xlc) */
  	  break;
  
***************
*** 1184,1190 ****
  	     the search if we have no valid frame yet or if the return
  	     address is not yet saved in the frame.  */
  	  if (fdata->frameless == 0
! 	      && (lr_reg == -1 || fdata->nosavedpc == 0))
  	    break;
  
  	  if (op == 0x4e800020		/* blr */
--- 1190,1196 ----
  	     the search if we have no valid frame yet or if the return
  	     address is not yet saved in the frame.  */
  	  if (fdata->frameless == 0
! 	      && (fdata->nosavedpc == 0))
  	    break;
  
  	  if (op == 0x4e800020		/* blr */
Index: gdb/thread.c
===================================================================
RCS file: /afs/slac/g/spear/cvsrep/tools/gdb/gdb/thread.c,v
retrieving revision 1.1.1.1
retrieving revision 1.3
diff -c -r1.1.1.1 -r1.3
*** gdb/thread.c	10 Nov 2004 18:25:50 -0000	1.1.1.1
--- gdb/thread.c	15 Mar 2005 02:46:27 -0000	1.3
***************
*** 431,438 ****
  	printf_filtered (" (%s)", extra_info);
        puts_filtered ("  ");
  
!       switch_to_thread (tp->ptid);
!       print_stack_frame (get_selected_frame (), 0, LOCATION);
      }
  
    switch_to_thread (current_ptid);
--- 431,446 ----
  	printf_filtered (" (%s)", extra_info);
        puts_filtered ("  ");
  
! 	  if ( current_target.to_shortname &&
! 	       strstr (current_target.to_shortname, "remote") )
! 		{
! 		  printf_filtered ("\n");
! 		}
! 	  else
! 		{
! 		  switch_to_thread (tp->ptid);
! 		  print_stack_frame (get_selected_frame (), 0, LOCATION);
! 		}
      }
  
    switch_to_thread (current_ptid);
Index: gdb/config/i386/i386-ssrlrtems.mt
===================================================================
RCS file: gdb/config/i386/i386-ssrlrtems.mt
diff -N gdb/config/i386/i386-ssrlrtems.mt
*** /dev/null	1 Jan 1970 00:00:00 -0000
--- gdb/config/i386/i386-ssrlrtems.mt	10 Nov 2004 23:12:42 -0000	1.1
***************
*** 0 ****
--- 1,4 ----
+ # Target: i386
+ include $(srcdir)/config/i386/i386.mt
+ TDEPFILES+= corelow.o core-regset.o i386-rtemscore.o
+ TM_FILE=tm-i386-ssrlrtems.h
Index: gdb/config/i386/tm-i386-ssrlrtems.h
===================================================================
RCS file: gdb/config/i386/tm-i386-ssrlrtems.h
diff -N gdb/config/i386/tm-i386-ssrlrtems.h
*** /dev/null	1 Jan 1970 00:00:00 -0000
--- gdb/config/i386/tm-i386-ssrlrtems.h	10 Nov 2004 23:12:42 -0000	1.1
***************
*** 0 ****
--- 1,41 ----
+ /* Target machine definitions for a generic m68k monitor/emulator
+    adding core file support.
+    
+    This file is part of GDB.
+ 
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+ 
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+ 
+    You should have received a copy of the GNU General Public License
+    along with this program; if not, write to the Free Software
+    Foundation, Inc., 59 Temple Place - Suite 330,
+    Boston, MA 02111-1307, USA.  */
+ 
+ /* GCC is probably the only compiler used on this configuration.  So
+    get this right even if the code which detects gcc2_compiled. is
+    still broken.  */
+ 
+ #ifndef TM_I386_SSRLRTEMS_H
+ #define TM_I386_SSRLRTEMS_H
+ 
+ #ifdef HAVE_STDINT_H
+ #include <stdint.h>
+ typedef uint32_t     rtemscore_uint32_t;
+ #else
+ typedef unsigned int rtemscore_uint32_t;
+ #endif
+ 
+ typedef rtemscore_uint32_t rtemscore_i386_gregset_t[16];
+ typedef unsigned rtemscore_i386_fpregset_t[0];
+ 
+ #define GDB_GREGSET_T rtemscore_i386_gregset_t
+ #define GDB_FPREGSET_T rtemscore_i386_fpregset_t
+ 
+ #endif /* defined TM_M68K_SSRLRTEMS_H */
Index: gdb/config/m68k/m68k-monitor-ssrlrtems.mt
===================================================================
RCS file: gdb/config/m68k/m68k-monitor-ssrlrtems.mt
diff -N gdb/config/m68k/m68k-monitor-ssrlrtems.mt
*** /dev/null	1 Jan 1970 00:00:00 -0000
--- gdb/config/m68k/m68k-monitor-ssrlrtems.mt	10 Nov 2004 23:12:44 -0000	1.1
***************
*** 0 ****
--- 1,4 ----
+ # Target: Motorola m68k embedded (EST emulator, rom68k and bug monitors)
+ include $(srcdir)/config/m68k/monitor.mt
+ TDEPFILES+= corelow.o core-regset.o m68k-rtemscore.o
+ TM_FILE= tm-m68k-monitor-ssrlrtems.h
Index: gdb/config/m68k/monitor.mt
===================================================================
RCS file: /afs/slac/g/spear/cvsrep/tools/gdb/gdb/config/m68k/monitor.mt,v
retrieving revision 1.1.1.1
retrieving revision 1.3
diff -c -r1.1.1.1 -r1.3
Index: gdb/config/m68k/tm-m68k-monitor-ssrlrtems.h
===================================================================
RCS file: gdb/config/m68k/tm-m68k-monitor-ssrlrtems.h
diff -N gdb/config/m68k/tm-m68k-monitor-ssrlrtems.h
*** /dev/null	1 Jan 1970 00:00:00 -0000
--- gdb/config/m68k/tm-m68k-monitor-ssrlrtems.h	10 Nov 2004 23:12:45 -0000	1.1
***************
*** 0 ****
--- 1,44 ----
+ /* Target machine definitions for a generic m68k monitor/emulator
+    adding core file support.
+    
+    This file is part of GDB.
+ 
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+ 
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+ 
+    You should have received a copy of the GNU General Public License
+    along with this program; if not, write to the Free Software
+    Foundation, Inc., 59 Temple Place - Suite 330,
+    Boston, MA 02111-1307, USA.  */
+ 
+ /* GCC is probably the only compiler used on this configuration.  So
+    get this right even if the code which detects gcc2_compiled. is
+    still broken.  */
+ 
+ #ifndef TM_M68K_SSRLRTEMS_H
+ #define TM_M68K_SSRLRTEMS_H
+ 
+ #include "m68k/tm-monitor.h"
+ 
+ #ifdef HAVE_STDINT_H
+ #include <stdint.h>
+ typedef uint32_t     rtemscore_uint32_t;
+ #else
+ typedef unsigned int rtemscore_uint32_t;
+ #endif
+ 
+ typedef rtemscore_uint32_t rtemscore_m68k_gregset_t[2*8+2];
+ typedef unsigned rtemscore_m68k_fpregset_t[0];
+ 
+ #define GDB_GREGSET_T rtemscore_m68k_gregset_t
+ #define GDB_FPREGSET_T rtemscore_m68k_fpregset_t
+ 
+ /* FIXME, should do GET_LONGJMP_TARGET for newlib.  */
+ #endif /* defined TM_M68K_SSRLRTEMS_H */
Index: gdb/config/m68k/tm-monitor.h
===================================================================
RCS file: /afs/slac/g/spear/cvsrep/tools/gdb/gdb/config/m68k/tm-monitor.h,v
retrieving revision 1.1.1.1
retrieving revision 1.3
diff -c -r1.1.1.1 -r1.3
Index: gdb/config/powerpc/ppc-eabi.mt
===================================================================
RCS file: /afs/slac/g/spear/cvsrep/tools/gdb/gdb/config/powerpc/ppc-eabi.mt,v
retrieving revision 1.1.1.1
retrieving revision 1.3
diff -c -r1.1.1.1 -r1.3
Index: gdb/config/powerpc/ppc-ssrlrtems.mt
===================================================================
RCS file: gdb/config/powerpc/ppc-ssrlrtems.mt
diff -N gdb/config/powerpc/ppc-ssrlrtems.mt
*** /dev/null	1 Jan 1970 00:00:00 -0000
--- gdb/config/powerpc/ppc-ssrlrtems.mt	12 Nov 2005 02:04:18 -0000	1.2
***************
*** 0 ****
--- 1,4 ----
+ # Target: PowerPC running eabi including simulator
+ include $(srcdir)/config/powerpc/ppc-sim.mt
+ TDEPFILES+= corelow.o core-regset.o ppc32-rtemscore.o
+ TM_FILE= tm-ppc-ssrlrtems.h
Index: gdb/config/powerpc/tm-ppc-eabi.h
===================================================================
RCS file: /afs/slac/g/spear/cvsrep/tools/gdb/gdb/config/powerpc/tm-ppc-eabi.h,v
retrieving revision 1.1.1.1
retrieving revision 1.3
diff -c -r1.1.1.1 -r1.3
Index: gdb/config/powerpc/tm-ppc-ssrlrtems.h
===================================================================
RCS file: gdb/config/powerpc/tm-ppc-ssrlrtems.h
diff -N gdb/config/powerpc/tm-ppc-ssrlrtems.h
*** /dev/null	1 Jan 1970 00:00:00 -0000
--- gdb/config/powerpc/tm-ppc-ssrlrtems.h	10 Nov 2004 23:12:47 -0000	1.1
***************
*** 0 ****
--- 1,41 ----
+ /* Macro definitions for Power PC running embedded ABI on rtems.
+    adding core file support
+ 
+    This file is part of GDB.
+ 
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+ 
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+ 
+    You should have received a copy of the GNU General Public License
+    along with this program; if not, write to the Free Software
+    Foundation, Inc., 59 Temple Place - Suite 330,
+    Boston, MA 02111-1307, USA.  */
+ 
+ #ifndef TM_PPC_SSRLRTEMS_H
+ #define TM_PPC_SSRLRTEMS_H
+ 
+ /* Use ppc/eabi definitions */
+ #include "powerpc/tm-ppc-eabi.h"
+ 
+ #ifdef HAVE_STDINT_H
+ #include <stdint.h>
+ typedef uint32_t     rtemscore_uint32_t;
+ #else
+ typedef unsigned int rtemscore_uint32_t;
+ #endif
+ 
+ typedef rtemscore_uint32_t rtemscore_ppc32_gregset_t[32+8]; /* we add 3 extensions: msr, dar and the vector */
+ typedef unsigned rtemscore_ppc32_fpregset_t[0];
+  
+ #define GDB_GREGSET_T rtemscore_ppc32_gregset_t
+ #define GDB_FPREGSET_T rtemscore_ppc32_fpregset_t
+ 
+ 
+ #endif /* TM_PPC_SSRLRTEMS_H */
Index: opcodes/po/da.gmo
===================================================================
RCS file: /afs/slac/g/spear/cvsrep/tools/gdb/opcodes/po/da.gmo,v
retrieving revision 1.1.1.1
retrieving revision 1.2
diff -c -r1.1.1.1 -r1.2
Binary files /tmp/cvsTkzQz4 and /tmp/cvspF0XPW differ
Index: sim/configure
===================================================================
RCS file: /afs/slac/g/spear/cvsrep/tools/gdb/sim/configure,v
retrieving revision 1.1.1.1
retrieving revision 1.2
diff -c -r1.1.1.1 -r1.2
*** sim/configure	10 Nov 2004 18:26:30 -0000	1.1.1.1
--- sim/configure	12 Nov 2005 02:05:40 -0000	1.2
***************
*** 1458,1464 ****
  	extra_subdirs="${extra_subdirs} testsuite"
  	;;
    powerpc*-*-eabi* | powerpc*-*-solaris* | powerpc*-*-sysv4* | \
!   powerpc*-*-elf* | powerpc*-*-linux* | powerpc*-*-netbsd* )
  	# The PowerPC simulator uses the GCC extension long long as well as
  	# ANSI prototypes, so don't enable it for random host compilers
  	# unless asked to.
--- 1458,1464 ----
  	extra_subdirs="${extra_subdirs} testsuite"
  	;;
    powerpc*-*-eabi* | powerpc*-*-solaris* | powerpc*-*-sysv4* | \
!   powerpc*-*-elf* | powerpc*-*-linux* | powerpc*-*-netbsd* | powerpc*-*-rtems* )
  	# The PowerPC simulator uses the GCC extension long long as well as
  	# ANSI prototypes, so don't enable it for random host compilers
  	# unless asked to.
Index: sim/ppc/hw_disk.c
===================================================================
RCS file: /afs/slac/g/spear/cvsrep/tools/gdb/sim/ppc/hw_disk.c,v
retrieving revision 1.1.1.1
retrieving revision 1.2
diff -c -r1.1.1.1 -r1.2
*** sim/ppc/hw_disk.c	10 Nov 2004 18:26:39 -0000	1.1.1.1
--- sim/ppc/hw_disk.c	10 Nov 2004 18:40:47 -0000	1.2
***************
*** 200,205 ****
--- 200,212 ----
  		hw_disk_device *disk,
  		const char *name)
  {
+ unsigned32 block_size;
+ 
+   if (device_find_property(me, "block-size"))
+     block_size = device_find_integer_property(me, "block-size");
+   else
+     block_size = 512;
+ 
    if (disk->image != NULL)
      fclose(disk->image);
    if (disk->name != NULL)
***************
*** 210,215 ****
--- 217,226 ----
      perror(device_name(me));
      device_error(me, "open %s failed\n", disk->name);
    }
+   if (setvbuf(disk->image, malloc(block_size), _IOFBF, block_size)) {
+ 	perror(device_name(me));
+     device_error(me, "setvbuf %s failed\n", disk->name);
+   }
  
    DTRACE(disk, ("image %s (%s)\n",
                  disk->name,
***************
*** 244,252 ****
  
    /* is it a RO device? */
    disk->read_only =
!     (strcmp(device_name(me), "disk") != 0
!      && strcmp(device_name(me), "floppy") != 0
!      && device_find_property(me, "read-only") == NULL);
  
    /* now open it */
    open_disk_image(me, disk, name);
--- 255,263 ----
  
    /* is it a RO device? */
    disk->read_only =
!     ( (strcmp(device_name(me), "disk") != 0
!        && strcmp(device_name(me), "floppy") != 0 )
!      || device_find_property(me, "read-only") != NULL);
  
    /* now open it */
    open_disk_image(me, disk, name);
